arbnode/hotshot_reader.go

```diff
@@ -0,0 +1,64 @@
+package arbnode
+
+import (
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/offchainlabs/nitro/arbos/espresso"
+	"github.com/offchainlabs/nitro/arbos/espresso/hotshot"
+)
+
+type HotShotReader struct {
+	HotShot hotshot.Hotshot
+}
+
+func NewHotShotReader(hotShotAddr common.Address, l1client bind.ContractBackend) (*HotShotReader, error) {
+	hotshot, err := hotshot.NewHotshot(hotShotAddr, l1client)
+	if err != nil {
+		return nil, err
+	}
+
+	return &HotShotReader{
+		HotShot: *hotshot,
+	}, nil
+}
+
+// L1HotShotCommitmentFromHeight returns a HotShot commitments to a sequencer block
+// This is used in the derivation pipeline to validate sequencer batches in Espresso mode
+func (h *HotShotReader) L1HotShotCommitmentFromHeight(blockHeight uint64) (*espresso.Commitment, error) {
+	var comm espresso.Commitment
+	// Check if the requested commitments are even available yet on L1.
+	contractBlockHeight, err := h.HotShot.HotshotCaller.BlockHeight(nil)
+	if err != nil {
+		return nil, err
+	}
+	if contractBlockHeight.Cmp(big.NewInt(int64(blockHeight))) < 0 {
+		return nil, fmt.Errorf("commitment at block height %d is unavailable (current contract block height %d)", blockHeight, contractBlockHeight)
+	}
```

Comment:
Maybe this check isn't necessary here. We did this check in the OP stack because we were fetching multiple commitments at a time, and didn't want to do the work of fetching the first few if they weren't all ready. But here we are only fetching one commitment anyway, and we can tell if it's not ready yet because it will be 0, so maybe we can skip this and save a call to the RPC.

Reply:
Added an issue: https://github.com/EspressoSystems/espresso-sequencer/issues/786